src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+}
+
+void
+LedgerState::commit(std::function<void()> f)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    if (mParent)
+    {
+        assert(!f);
+        mergeStateIntoParent();
+        mergeHeaderIntoParent();
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        soci::transaction sqlTx(mRoot->getDatabase().getSession());
+        try
+        {
+            mergeStateIntoRoot();
+            mergeHeaderIntoRoot();
+            if (f)
+            {
+                f();
+            }
+            sqlTx.commit();
+        }
+        catch (...)
+        {
+            mRoot->getCache().clear();
+            throw;
+        }
+
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+void
+LedgerState::mergeStateIntoParent()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        auto const& entry = state.second->ignoreInvalid().entry();
+
+        auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
+            ler->invalidate();
+            mParent->mState[state.first] = ler;
+        }
+        else
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            StateEntry ler(new LedgerEntryReference(entry, previous));
+            ler->invalidate();
+            iter->second = ler;
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoParent()
+{
+    if (mHeader)
+    {
+        auto const& header = mHeader->ignoreInvalid().header();
+        auto const& previous = mHeader->ignoreInvalid().previousHeader();
+        StateHeader lhr(new LedgerHeaderReference(header, previous));
+        lhr->invalidate();
+        mParent->mHeader = lhr;
+    }
+}
+
+void
+LedgerState::mergeStateIntoRoot()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        if (state.second->ignoreInvalid().entry())
+        {
+            storeInDatabase(state.second);
+        }
+        else
+        {
+            deleteFromDatabase(state.second);
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoRoot()
+{
+    if (mHeader)
+    {
+        storeHeaderInDatabase();
+        mRoot->setPreviousHeader(mHeader->ignoreInvalid().header());
+    }
+}
+
+void
+LedgerState::rollback()
+{
+    assert(mParent || mRoot);
+
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    if (mParent)
+    {
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+LedgerState::StateEntry
+LedgerState::create(LedgerEntry const& entry)
+{
+    assert(mParent || mRoot);
```

Comment:
this set of assets appears everywhere, maybe promote it to a method?

Reply:
I considered this also (it was actually that way in the first draft of the code) but I wasn't sure it would actually be easier to follow in the future if that was a method. If the consensus is that it would be easier that way, I'm not opposed to changing it back.